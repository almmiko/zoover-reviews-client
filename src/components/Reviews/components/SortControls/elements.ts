import styled from 'styled-components';

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
`;

export const Title = styled.div`
  color: #757575;
  margin-right: 10px;
`;
