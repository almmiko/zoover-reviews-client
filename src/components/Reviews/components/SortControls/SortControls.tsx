import React from 'react';
import { Button, Icon } from 'antd';
import { Wrapper, Title } from './elements';

type State = {
  travelDateOrder: string,
  reviewDateOrder: string,
};

enum Order {
  ascending = 'asc',
  descending = 'desc',
}

type Props = {
  loading?: boolean,
  onSort: (sortBy: string, order: string) => void,
};

class SortControls extends React.Component<Props, State> {

  state = {
    travelDateOrder: Order.descending,
    reviewDateOrder: Order.descending
  };

  getIcon = (sortType: string) => {
    const order = this.state[sortType];

    switch (order) {
      case Order.ascending:
        return 'caret-down';
      case Order.descending:
        return 'caret-up';
      default:
        return 'caret-up'
    }
  };

  handleTravelDateOrder = () => {
    const { travelDateOrder } = this.state;

    this.setState({
      travelDateOrder: travelDateOrder === Order.ascending ? Order.descending : Order.ascending,
    }, () => {
      this.props.onSort('travelDate', this.state.travelDateOrder);
    });

  };
  // let's keep this fn and handleTravelDateOrder for better logic splitting.
  handleReviewDateOrder = () => {
    const { reviewDateOrder } = this.state;

    this.setState({
      reviewDateOrder: reviewDateOrder === Order.ascending ? Order.descending : Order.ascending,
    }, () => {
      this.props.onSort('entryDate', this.state.reviewDateOrder);
    });

  };

  render() {
    return (
      <Wrapper>
        <Title>
          Sort by:
        </Title>
        <Button id={'travelDateOrderButton'} type="default" onClick={this.handleTravelDateOrder}>
          Travel Date
          <Icon type={(() => this.getIcon('travelDateOrder'))()} />
        </Button>
        <Button type="default" onClick={this.handleReviewDateOrder}>
          Review Date
          <Icon type={(() => this.getIcon('reviewDateOrder'))()} />
        </Button>
      </Wrapper>
    );
  }
}

export default SortControls;
