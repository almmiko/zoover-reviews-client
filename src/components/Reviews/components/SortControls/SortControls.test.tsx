import React from 'react';
import SortControls from './SortControls';
import { mount, shallow } from 'enzyme';
import { Button } from 'antd';

it('renders correctly', () => {
  const onSort = jest.fn();
  const tree = shallow(<SortControls onSort={onSort} />);
  expect(tree).toMatchSnapshot();
});

it('should toggle state on onSort callback', () => {
  const onSort = jest.fn();
  const tree = mount(<SortControls onSort={onSort} />);

  tree.find(Button).children().first().simulate('click');
  tree.find(Button).children().first().simulate('click');

  expect(onSort).toBeCalledWith('travelDate', 'asc');
  expect(onSort).toHaveBeenLastCalledWith('travelDate', 'desc');
  expect(onSort).toBeCalledTimes(2);
});
