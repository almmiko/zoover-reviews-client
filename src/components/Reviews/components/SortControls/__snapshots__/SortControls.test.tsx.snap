// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<styled.div>
  <styled.div>
    Sort by:
  </styled.div>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    id="travelDateOrderButton"
    loading={false}
    onClick={[Function]}
    type="default"
  >
    Travel Date
    <Icon
      type="caret-up"
    />
  </Button>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[Function]}
    type="default"
  >
    Review Date
    <Icon
      type="caret-up"
    />
  </Button>
</styled.div>
`;
